Coupons: Hide limit usage to X Items when discount type is fixed cart

ThomazFB opened the pull request:
Summary
==========
Fix issue #6863 by setting a rule inside the `CouponRestriction` view and viewModel to only display the `limit usage to X items` row when the Coupon discount type is different from `fixed cart`.

Screenshots
==========
### Before
https://user-images.githubusercontent.com/5920403/170117633-d1a45346-2f4d-4230-9df8-b984ac7d49ce.mp4

### After
https://user-images.githubusercontent.com/5920403/170117670-c85a7417-645d-4db7-a535-36299fd63068.mp4

How to Test
==========
### Scenario 1
1. Make sure the Coupons experimental toggle is activated
2. Go to the Coupon list, select a Coupon configured with the fixed cart discount type
3. Inside the Coupon details, select the Edit coupon in the action menu
4. Inside the Edit Coupon view, select the `usage restrictions` option
5. Verify that the `limit usage to X items` is **NOT** displayed

### Scenario 2
1. Make sure the Coupons experimental toggle is activated
2. Go to the Coupon list, select a Coupon configured with **any discount type different from fixed cart**
3. Inside the Coupon details, select the Edit coupon in the action menu
4. Inside the Edit Coupon view, select the `usage restrictions` option
5. Verify that the `limit usage to X items` is displayed

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Adjust usage to X items limit field to be kept in a dedicated Vstack with divider

ThomazFB committed:
Define shouldDisplayLimitUsageToXItemsRow property

ThomazFB committed:
Fix init not setting shouldDisplayLimitUsageToXItemsRow property

ThomazFB committed:
Fix shouldDisplayLimitUsageToXItemsRow value constraint definition

ThomazFB committed:
Set limitUsageToXItems row rendering constraint to be defined by the CouponRestrictionsViewModel

ThomazFB committed:
Add unit test to enforce shouldDisplayLimitUsageToXItemsRow rule

ThomazFB committed:
Improve CouponRestrictionsViewModelTests

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/UsageDetails/CouponRestrictionsViewModel.swift:
```diff
@@ -9,6 +9,8 @@ final class CouponRestrictionsViewModel: ObservableObject {
 
     let currencySymbol: String
 
+    let shouldDisplayLimitUsageToXItemsRow: Bool
```
I think this property should be a computed variable based on `discountType`. This will be important since I'm adding the option to update the discount type in #6935 - so this property will no longer be constant. Using a computed variable makes sure that `shouldDisplayLimitUsageToXItemsRow` will always be updated when the discount type is updated.

```suggestion
    var shouldDisplayLimitUsageToXItemsRow: Bool {
        discountType != .fixedCart
    }
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,20 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+
+        // Then
+        XCTAssertTrue(viewModel.shouldDisplayLimitUsageToXItemsRow)
+    }
```
Thanks for adding these tests!

itsmeichigo reviewed (commented):
Thanks, Thomaz for the fix! This is working for the current implementation of the Edit Coupon screen 👍 

But a new update is coming with the addition of the option to update discount type in #6935, so we will need to update the visibility of "Limit usage to X Items" accordingly (more details in the comment below).

Another problem I found is that if we set a value to this field when a coupon is of another discount type, and then update the discount type to fixed cart, the field is still visible on the Coupon Details screen. We should probably check to hide this field on this screen also:

<img src="https://user-images.githubusercontent.com/5533851/170176811-c75c0101-8123-4457-9c36-c0128a2408cf.png" width=320 />

ThomazFB committed:
Define shouldDisplayLimitUsageToXItems property in CouponDetailsViewModel

ThomazFB committed:
Configure CouponDetails to use shouldDisplayLimitUsageToXItems

ThomazFB commented:
Thanks @itsmeichigo for your great review! And nice catch with the Coupon details part! I'll fix it later today 🙇 

I have just one question regarding changing the `shouldDisplayLimitUsageToXItemsRow` to computed value property and the issue we're trying to avoid: 

I see that the `CouponRestrictionsViewModel` is stored inside the `AddEditCouponViewModel`, and when we update the discount type in the `AddEditCoupon` view, this change should be reflected inside the `CouponRestrictionsViewModel` too, but looking how it is right now, the `existingCoupon` is only passed there when we initialize the ViewModel, and that ViewModel doesn't store the Coupon, only extracts some specific data. 

So, correct me if I'm wrong, but it looks like the `CouponRestrictionsViewModel` currently doesn't have a way to get this updated discount type that we can use to allow evaluating `shouldDisplayLimitUsageToXItemsRow` any time as a computed value. Maybe we should update this ViewModel inside the `AddEditCoupon` whenever we call the `CouponRestrictions` view?

itsmeichigo commented:
> but it looks like the `CouponRestrictionsViewModel` currently doesn't have a way to get this updated discount type that we can use to allow evaluating `shouldDisplayLimitUsageToXItemsRow` any time as a computed value.

Great question @ThomazFB! I admit that my suggestion wasn't correct earlier 😅 I was thinking that `shouldDisplayLimitUsageToXItemsRow` was created in `AddEditCouponViewModel` and didn't realize it was in `CouponRestrictionsViewModel`, hence the suggestion.

> Maybe we should update this ViewModel inside the AddEditCoupon whenever we call the CouponRestrictions view?

I'm thinking of something close to this: let's add a property observer to `AddEditCouponViewModel`'s discount type (using `didSet`). So whenever the discount type is updated, we update `shouldDisplayLimitUsageToXItemsRow` on the `CouponRestrictionsViewModel` from the `AddEditCouponViewModel` accordingly. This should be doable since the `AddEditCouponViewModel` keeps a reference to the accompanying `CouponRestrictionsViewModel`. 

How does that sound?

ThomazFB committed:
Adapt discountType to be a published variable

ThomazFB committed:
Adjust shouldDisplayLimitUsageToXItemsRow to be changed from AddEditCouponViewModel

ThomazFB committed:
Adjust AddEditCouponViewModel to notify CouponRestrictionViewModel when the discountType changes

ThomazFB committed:
Improve CouponRestrictionsViewModelTests

ThomazFB committed:
Improve CouponDetailsViewModelTests

ThomazFB commented:
@itsmeichigo thanks for the suggestions and discussion! I've addressed the points we discussed and both Coupon details and the Coupon Restrictions are now updated to be aware of discount type changes. I modified the discountType inside the `AddEditCouponViewModel` following your changes in https://github.com/woocommerce/woocommerce-ios/pull/6935 to avoid a complicated merge between both this PR and yours and applied the `didSet` idea. Let me know what you think!

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
     }
+
+    func test_shouldDisplayLimitUsageToXItemsRow_is_false_when_fixed_cart_discount_type_is_set() {
+        let sampleCoupon = Coupon.fake().copy(discountType: .fixedCart)
```
Nit: We should add clear comments for the condition and expectation of each test following Given/When/Then pattern.
```suggestion
        // Given
        let sampleCoupon = Coupon.fake().copy(discountType: .fixedCart)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+    }
+
+    func test_shouldDisplayLimitUsageToXItemsRow_is_true_when_fixed_cart_discount_type_is_NOT_set() {
+        let sampleCoupon = Coupon.fake().copy(discountType: .percent)
```
Nit: Same comment as above about the missing comment.
```suggestion
        // Given
        let sampleCoupon = Coupon.fake().copy(discountType: .percent)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+    }
+
+    func test_shouldDisplayLimitUsageToXItemsRow_is_false_when_discount_type_is_changed_to_fixed_cart() {
+        let sampleCoupon = Coupon.fake().copy(discountType: .percent)
```
Nit:
```suggestion
        // Given
        let sampleCoupon = Coupon.fake().copy(discountType: .percent)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+        // Then
+        XCTAssertTrue(viewModel.shouldDisplayLimitUsageToXItemsRow)
+
+        viewModel.onDiscountTypeChanged(discountType: .fixedCart)
```
Nit:
```suggestion
        // When
        viewModel.onDiscountTypeChanged(discountType: .fixedCart)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+
+        viewModel.onDiscountTypeChanged(discountType: .fixedCart)
+
+        XCTAssertFalse(viewModel.shouldDisplayLimitUsageToXItemsRow)
```
Nit:
```suggestion
        // Then
        XCTAssertFalse(viewModel.shouldDisplayLimitUsageToXItemsRow)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+    }
+
+    func test_shouldDisplayLimitUsageToXItemsRow_is_true_when_discount_type_is_changed_from_fixed_cart() {
+        let sampleCoupon = Coupon.fake().copy(discountType: .fixedCart)
```
Nit:
```suggestion
        // Given
        let sampleCoupon = Coupon.fake().copy(discountType: .fixedCart)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+        // Then
+        XCTAssertFalse(viewModel.shouldDisplayLimitUsageToXItemsRow)
+
+        viewModel.onDiscountTypeChanged(discountType: .percent)
```
Nit:
```suggestion
        // When
        viewModel.onDiscountTypeChanged(discountType: .percent)
```

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift:
```diff
@@ -68,4 +68,44 @@ final class CouponRestrictionsViewModelTests: XCTestCase {
+
+        viewModel.onDiscountTypeChanged(discountType: .percent)
+
+        XCTAssertTrue(viewModel.shouldDisplayLimitUsageToXItemsRow)
```
Nit: 
```suggestion
        // Then
        XCTAssertTrue(viewModel.shouldDisplayLimitUsageToXItemsRow)
```

itsmeichigo reviewed (approved):
This works perfectly now 🎉 

I left a few nitpicking comments regarding the unit tests, please feel free to merge this afterward if you fix those issues. Thanks!

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

ThomazFB committed:
Update WooCommerce/WooCommerceTests/ViewRelated/Coupons/CouponRestrictionsViewModelTests.swift

Co-authored-by: Huong Do <hi.huongdo@gmail.com>
